File: gestaltd/internal/providerhost/cache_server.go
package providerhost

import (
	"context"
	"fmt"

	proto "github.com/valon-technologies/gestalt/sdk/go/gen/v1"
	corecache "github.com/valon-technologies/gestalt/server/core/cache"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
)

type cacheServer struct {
	proto.UnimplementedCacheServer
	cache  corecache.Cache
	plugin string
}

func NewCacheServer(cache corecache.Cache, pluginName string) proto.CacheServer {
	return &cacheServer{cache: cache, plugin: pluginName}
}

func (s *cacheServer) Get(ctx context.Context, req *proto.CacheGetRequest) (*proto.CacheGetResponse, error) {
	value, found, err := s.cache.Get(ctx, s.key(req.GetKey()))
	if err != nil {
		return nil, err
	}
	return &proto.CacheGetResponse{Found: found, Value: value}, nil
}

func (s *cacheServer) GetMany(ctx context.Context, req *proto.CacheGetManyRequest) (*proto.CacheGetManyResponse, error) {
	keys := req.GetKeys()
	namespaced := make([]string, 0, len(keys))
	for _, key := range keys {
		namespaced = append(namespaced, s.key(key))
	}
	values, err := s.cache.GetMany(ctx, namespaced)
	if err != nil {
		return nil, err
	}
	entries := make([]*proto.CacheResult, 0, len(keys))
	for _, key := range keys {
		entry := &proto.CacheResult{Key: key}
		if value, ok := values[s.key(key)]; ok {
			entry.Found = true
			entry.Value = value
		}
		entries = append(entries, entry)
	}
	return &proto.CacheGetManyResponse{Entries: entries}, nil
}

func (s *cacheServer) Set(ctx context.Context, req *proto.CacheSetRequest) (*emptypb.Empty, error) {
	ttl, err := ttlFromProto(req.GetTtl())
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, err.Error())
	}
	if err := s.cache.Set(ctx, s.key(req.GetKey()), req.GetValue(), corecache.SetOptions{TTL: ttl}); err != nil {
		return nil, err
	}
	return &emptypb.Empty{}, nil
}

func (s *cacheServer) SetMany(ctx context.Context, req *proto.CacheSetManyRequest) (*emptypb.Empty, error) {
	ttl, err := ttlFromProto(req.GetTtl())
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, err.Error())
	}
	entries := make([]corecache.Entry, 0, len(req.GetEntries()))
	for _, entry := range req.GetEntries() {
		entries = append(entries, corecache.Entry{
			Key:   s.key(entry.GetKey()),
			Value: entry.GetValue(),
		})
	}
	if err := s.cache.SetMany(ctx, entries, corecache.SetOptions{TTL: ttl}); err != nil {
		return nil, err
	}
	return &emptypb.Empty{}, nil
}

func (s *cacheServer) Delete(ctx context.Context, req *proto.CacheDeleteRequest) (*proto.CacheDeleteResponse, error) {
	deleted, err := s.cache.Delete(ctx, s.key(req.GetKey()))
	if err != nil {
		return nil, err
	}
	return &proto.CacheDeleteResponse{Deleted: deleted}, nil
}

func (s *cacheServer) DeleteMany(ctx context.Context, req *proto.CacheDeleteManyRequest) (*proto.CacheDeleteManyResponse, error) {
	keys := make([]string, 0, len(req.GetKeys()))
	for _, key := range req.GetKeys() {
		keys = append(keys, s.key(key))
	}
	deleted, err := s.cache.DeleteMany(ctx, keys)
	if err != nil {
		return nil, err
	}
	return &proto.CacheDeleteManyResponse{Deleted: deleted}, nil
}

func (s *cacheServer) Touch(ctx context.Context, req *proto.CacheTouchRequest) (*proto.CacheTouchResponse, error) {
	ttl, err := ttlFromProto(req.GetTtl())
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, err.Error())
	}
	touched, err := s.cache.Touch(ctx, s.key(req.GetKey()), ttl)
	if err != nil {
		return nil, err
	}
	return &proto.CacheTouchResponse{Touched: touched}, nil
}

func (s *cacheServer) key(key string) string {
	if s.plugin == "" {
		return key
	}
	return fmt.Sprintf("%s:%s", s.plugin, key)
}

File: sdk/rust/src/cache_server.rs
use std::sync::Arc;
use std::time::Duration;

use tonic::{Request as GrpcRequest, Response as GrpcResponse, Status};

use crate::cache::{CacheEntry, CacheProvider, CacheSetOptions};
use crate::generated::v1::cache_server::Cache as CacheGrpc;
use crate::generated::v1::{
    CacheDeleteManyRequest, CacheDeleteManyResponse, CacheDeleteRequest, CacheDeleteResponse,
    CacheGetManyRequest, CacheGetManyResponse, CacheGetRequest, CacheGetResponse, CacheResult,
    CacheSetManyRequest, CacheSetRequest, CacheTouchRequest, CacheTouchResponse,
};
use crate::rpc_status::rpc_status;

#[derive(Clone)]
pub struct CacheRpcServer<P> {
    cache: Arc<P>,
}

impl<P> CacheRpcServer<P> {
    pub fn new(cache: Arc<P>) -> Self {
        Self { cache }
    }
}

#[tonic::async_trait]
impl<P> CacheGrpc for CacheRpcServer<P>
where
    P: CacheProvider,
{
    async fn get(
        &self,
        request: GrpcRequest<CacheGetRequest>,
    ) -> std::result::Result<GrpcResponse<CacheGetResponse>, Status> {
        let request = request.into_inner();
        let value = self
            .cache
            .get(&request.key)
            .await
            .map_err(|error| rpc_status("get cache entry", error))?;
        match value {
            Some(value) => Ok(GrpcResponse::new(CacheGetResponse { found: true, value })),
            None => Ok(GrpcResponse::new(CacheGetResponse {
                found: false,
                value: Vec::new(),
            })),
        }
    }

    async fn get_many(
        &self,
        request: GrpcRequest<CacheGetManyRequest>,
    ) -> std::result::Result<GrpcResponse<CacheGetManyResponse>, Status> {
        let request = request.into_inner();
        let values = self
            .cache
            .get_many(&request.keys)
            .await
            .map_err(|error| rpc_status("get many cache entries", error))?;
        let entries = request
            .keys
            .into_iter()
            .map(|key| match values.get(&key) {
                Some(value) => CacheResult {
                    key,
                    found: true,
                    value: value.clone(),
                },
                None => CacheResult {
                    key,
                    found: false,
                    value: Vec::new(),
                },
            })
            .collect();
        Ok(GrpcResponse::new(CacheGetManyResponse { entries }))
    }

    async fn set(
        &self,
        request: GrpcRequest<CacheSetRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        let request = request.into_inner();
        self.cache
            .set(
                &request.key,
                &request.value,
                CacheSetOptions {
                    ttl: duration_from_proto(request.ttl)?,
                },
            )
            .await
            .map_err(|error| rpc_status("set cache entry", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn set_many(
        &self,
        request: GrpcRequest<CacheSetManyRequest>,
    ) -> std::result::Result<GrpcResponse<()>, Status> {
        let request = request.into_inner();
        let entries: Vec<CacheEntry> = request
            .entries
            .into_iter()
            .map(|entry| CacheEntry {
                key: entry.key,
                value: entry.value,
            })
            .collect();
        self.cache
            .set_many(
                &entries,
                CacheSetOptions {
                    ttl: duration_from_proto(request.ttl)?,
                },
            )
            .await
            .map_err(|error| rpc_status("set many cache entries", error))?;
        Ok(GrpcResponse::new(()))
    }

    async fn delete(
        &self,
        request: GrpcRequest<CacheDeleteRequest>,
    ) -> std::result::Result<GrpcResponse<CacheDeleteResponse>, Status> {
        let request = request.into_inner();
        let deleted = self
            .cache
            .delete(&request.key)
            .await
            .map_err(|error| rpc_status("delete cache entry", error))?;
        Ok(GrpcResponse::new(CacheDeleteResponse { deleted }))
    }

    async fn delete_many(
        &self,
        request: GrpcRequest<CacheDeleteManyRequest>,
    ) -> std::result::Result<GrpcResponse<CacheDeleteManyResponse>, Status> {
        let deleted = self
            .cache
            .delete_many(&request.into_inner().keys)
            .await
            .map_err(|error| rpc_status("delete many cache entries", error))?;
        Ok(GrpcResponse::new(CacheDeleteManyResponse { deleted }))
    }

    async fn touch(
        &self,
        request: GrpcRequest<CacheTouchRequest>,
    ) -> std::result::Result<GrpcResponse<CacheTouchResponse>, Status> {
        let request = request.into_inner();
        let ttl = duration_from_proto(request.ttl)?.unwrap_or_default();
        let touched = self
            .cache
            .touch(&request.key, ttl)
            .await
            .map_err(|error| rpc_status("touch cache entry", error))?;
        Ok(GrpcResponse::new(CacheTouchResponse { touched }))
    }
}

fn duration_from_proto(
    ttl: Option<prost_types::Duration>,
) -> std::result::Result<Option<Duration>, Status> {
    let Some(ttl) = ttl else {
        return Ok(None);
    };
    if ttl.seconds < 0 {
        return Err(Status::invalid_argument("ttl must be non-negative"));
    }
    if ttl.nanos < 0 || ttl.nanos >= 1_000_000_000 {
        return Err(Status::invalid_argument("ttl nanos must be in [0, 1e9)"));
    }
    let seconds =
        u64::try_from(ttl.seconds).map_err(|_| Status::invalid_argument("ttl is too large"))?;
    Ok(Some(Duration::new(seconds, ttl.nanos as u32)))
}
